### PR TITLE
Add FaaS fields to intake

### LINKED ISSUE #5753: Extend Intake to accept FaaS / Serverless context fields
With [adding AWS Lambda support to our agents](https://github.com/elastic/apm/pull/470) we need additional context in the intake API that accepts function as a service (faas)-specific fields.

Fields to be added:

Field | Example | Metadata | Transaction | Description
---   | ---   | --- | --- | ---
`service.id` | e.g. `arn:aws:lambda:us-west-2:123456789012:function:my-function` | yes | yes | Add `service.id`(available in ECS) to the intake model. Value for Lambda: The ARN of the function without alias suffix. 
`faas.trigger.type` | `http` | no | yes | The trigger type. 
`faas.execution` | `af9aa4-a6...` | no | yes | The AWS request ID of the function invocation 
`faas.coldstart` | `true` / `false` | no | yes | Boolean value indicating whether a Lambda function invocation was a cold start or not.
`faas.trigger.reuqest_id` | `af9aa4-a6...` | no | yes | The ID of the origin trigger request
`cloud.origin.*` | - |  no  |  yes | Allow to self-nest all cloud fields under `cloud.origin` to specify trigger cloud data. See https://github.com/elastic/ecs/pull/1542
`service.origin.*` | - |  no  |  yes | Allow to self-nest all service fields under `service.origin` to specify trigger service data. See https://github.com/elastic/ecs/pull/1542

Names and fields are inspired by the [OpenTelemetry specification for functions as a service](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/semantic_conventions/faas.md).

There is a [PR for starting the RFC process](https://github.com/elastic/ecs/pull/1518) to add these fields also to ECS.

### PR DESCRIPTION
**Pull request by stuartnelson3**
## Motivation/summary

Extend intake to include FaaS fields

## Checklist

- [x] Update [CHANGELOG.asciidoc](https://github.com/elastic/apm-server/blob/master/CHANGELOG.asciidoc)
- [x] Documentation has been updated

## How to test these changes

1. Ingest some documents that include FaaS fields
2. Confirm they are properly indexed

## Related issues

closes #5753

### CODE REVIEW

File: model/modeldecoder/v2/model.go [hunk 1]
```diff
@@ -93,6 +98,51 @@ type context struct {
...
+	// The cloud project identifier.
+	ProjectID nullable.String `json:"project.id"`
+	// Name of the cloud provider.
+	Provider nullable.String `json:"provider"`
+	// Region in which this host, resource, or service is located.
+	Region nullable.String `json:"region"`
+	// The cloud service name is intended to distinguish services running
+	// on different platforms within a provider.
+	ServiceName nullable.String `json:"service.name"`
+}
```
Comment (axw) [inline model/modeldecoder/v2/model.go:144]: I'm not sure that all of these make sense. For example, would we ever know the machine type or instance ID of the origin? https://github.com/elastic/apm/pull/470/files proposes just the following:

- `cloud.origin.provider`
- `cloud.origin.region`
- `cloud.origin.service.name`
- `cloud.origin.account.id`

I would prefer to start out with only those that we need, and expand as necessary.

File: model/modeldecoder/v2/model.go [hunk 2]
```diff
@@ -207,20 +257,41 @@ type contextService struct {
...
+	// Origin contains the self-nested field groups for service.
+	Origin contextServiceOrigin `json:"origin"`
 	// Runtime holds information about the language runtime running the
 	// monitored service
 	Runtime contextServiceRuntime `json:"runtime"`
 	// Version of the monitored service.
 	Version nullable.String `json:"version" validate:"maxLength=1024"`
 }
 
+type contextServiceOrigin struct {
```
Comment (axw) [inline model/modeldecoder/v2/model.go:278]: Similar to above, https://github.com/elastic/apm/pull/470/files proposes:

 - `service.origin.name`
 - `service.origin.version`
 - `service.origin.id`

File: model/modeldecoder/v2/model.go (hunk 1 above)
Comment (stuartnelson3) [reply model/modeldecoder/v2/model.go:144]: That's fine by me. When I asked @AlexanderWert he said to limit it to all service.* and cloud.* fields available in ECS, so I added what was there.

Comment (AlexanderWert) [reply model/modeldecoder/v2/model.go:144]: Sorry, my fault, was not aware we are already limiting the `cloud.*` and `service.*` intake fields to a subset of ECS. 
I'm also fine with @axw 's  proposal for `cloud.origin.*` and `service.origin.*`.

From the [AWS Lambda instrumentation spec PR](https://github.com/elastic/apm/blob/b98e151e4379f17754a393cb080875f0ca36e1b4/specs/agents/tracing-instrumentation-aws-lambda.md) the following *.origin.* fields are relevant:

- `service.origin.name`
- `service.origin.version`
- `service.origin.id`
- `cloud.origin.provider`
- `cloud.origin.region`
- `cloud.origin.service.name`
- `cloud.origin.account.id`

**Comment by stuartnelson3** [comment]
I'm still going to add a sample transaction containing these new fields to `testdata/intake-v2/transaction.ndjson`, but I'm not planning on adding any additional implementation, so I'm marking this as "ready for review".

File: model/modeldecoder/v2/model.go [hunk 3]
```diff
@@ -208,20 +242,33 @@ type contextService struct {
 	// Framework holds information about the framework used in the
 	// monitored service.
 	Framework contextServiceFramework `json:"framework"`
+	// ID holds a unique identifier for the event.
```
Comment (axw) [inline model/modeldecoder/v2/model.go:245]:
```suggestion
	// ID holds a unique identifier for the service.
```

File: model/service.go [hunk 4]
```diff
@@ -30,6 +31,16 @@ type Service struct {
 	Runtime     Runtime
 	Framework   Framework
 	Node        ServiceNode
+
+	Origin *ServiceOrigin
+}
+
+// ServiceOrigin holds self nested information related to the monitored
+// service.
```
Comment (axw) [inline model/service.go:39]: Not sure about this. Can we say it's information related to the originating service? e.g.

`// ServiceOrigin holds information about the service that originated a transaction.`

File: model/transaction/_meta/fields.yml [hunk 5]
```diff
@@ -221,6 +254,50 @@
               Version of the framework used.
             overwrite: true
 
+        - name: origin
```
Comment (axw) [inline model/transaction/_meta/fields.yml:257]: Can we limit this to only the fields that we'll set? i.e. id, name, and version

File: model/transaction/_meta/fields.yml [hunk 6]
```diff
@@ -810,6 +887,69 @@
           description: Cloud instance/machine type
           example: t2.medium
           overwrite: true
+      - name: origin
```
Comment (axw) [inline model/transaction/_meta/fields.yml:890]: same as above, just include the fields we might set

File: model/modeldecoder/v2/decoder.go [hunk 7]
```diff
@@ -698,6 +729,20 @@ func mapToServiceModel(from contextService, out *model.Service) {
 	if from.Version.IsSet() {
 		out.Version = from.Version.Val
 	}
+	outOrigin := out.Origin
+	if outOrigin == nil {
+		outOrigin = &model.ServiceOrigin{}
+	}
```
Comment (axw) [inline model/modeldecoder/v2/decoder.go:735]: we should probably only set this `if from.Origin.IsSet()`
same for cloud and faas

**Review by axw — COMMENTED** [review]
Looks good! I'd just like to trim back the field definitions to only those that we'll set. That + testdata, and I think this is good to go.

File: model/transaction/_meta/fields.yml (hunk 5 above)
Comment (stuartnelson3) [reply model/transaction/_meta/fields.yml:257]: oops! yes, I forgot to change this when I updated the code

**Comment by stuartnelson3** [comment]
I added the `cloud` and `service` sections under `context`. I assume `service.origin` will be under `context` as `service` is already there, and I guessed `cloud` will go there as well.

<details>
<summary>indented tx.json</summary>

```json
{
    "transaction": {
        "name": "july-2021-delete-after-july-31",
        "type": "lambda",
        "result": "success",
        "id": "142e61450efb8574",
        "trace_id": "eb56529a1f461c5e7e2f66ecb075e983",
        "subtype": null,
        "action": null,
        "duration": 38.853,
        "timestamp": 1631736666365048,
        "sampled": true,
        "context": {
          "cloud": {
            "origin": {
              "account": { "id": "abc123" },
              "provider": "aws",
              "region": "us-east-1",
              "service": { "name": "serviceName" }
            }
          },
          "service": {
            "origin": {
              "id": "abc123",
              "name": "service-name",
              "version": "1.0"
            }
          },
          "user": {},
          "tags": {},
          "custom": {

          }
        },
        "sync": true,
        "span_count": {
            "started": 0
        },
        "outcome": "unknown",
        "faas": {
            "coldstart": false,
            "execution": "2e13b309-23e1-417f-8bf7-074fc96bc683",
            "trigger": {
                "request_id": "FuH2Cir_vHcEMUA=",
                "type": "http"
            }
        },
        "sample_rate": 1
    }
}
```
</details>

**Review by axw — APPROVED** [review]
> I added the cloud and service sections under context. I assume service.origin will be under context as service is already there, and I guessed cloud will go there as well.

Yes, I think that's accurate

**Pull request merged**